cast Distribution Inputs to floatX instead of tensor_variable

hvasbath opened the pull request:
Removes slowdown again:
https://github.com/pymc-devs/pymc3/issues/1824
Running the test from that thread:
[v8, v7, v6, v5, v4, v3, v2, v1]
('time for initialising population', 1.4203970432281494) previously 20s +

Also fixes AttributeError in multivariate.py line 250.

Was not sure about the Mixture Distribution as there are a lot of theano and numpy operators mixed in the random method generation. Someone having insight should fix that one ...

hvasbath committed:
cast Distribution Inputs to floatX instead of tensor_variable

hvasbath committed:
cast Distribution Inputs to floatX instead of tensor_variable

junpenglao commented:
Could you test if it pass test in https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions.py#L827-L838 locally? The _repr_latex is using the `Name` from Theano tensor.

hvasbath commented:
Yep that one fails! How to deal with that? You could simply use the Distribution name and attribute names?

junpenglao commented:
I think we should change the function https://github.com/pymc-devs/pymc3/blob/master/pymc3/util.py#L79

junpenglao commented:
This might work:

```python
def get_variable_name(variable):
     """Returns the variable data type if it is a constant, otherwise
     returns the argument name.
     """
     try:
        name = variable.name
     except AttributeError:
        name = None
     if name is None:
        if hasattr(variable, 'get_parents'):
            try:
                names = [get_variable_name(item) for item in variable.get_parents()[0].inputs]
                return 'f(%s)' % ','.join([n for n in names if isinstance(n, str)]) 
            except IndexError:
                pass
        try:
            value = variable.eval()
        except AttributeError:
            value = variable
        if not value.shape:
            return asscalar(value)
        return 'array'
    return name
```
for some reason I cannot check out your branch. Also any idea why Travis is not building @twiecki ?

ColCarroll commented on pymc3/distributions/multivariate.py:
```diff
@@ -247,7 +247,7 @@ def _repr_latex_(self, name=None, dist=None):
         mu = dist.mu
         try:
             cov = dist.cov
-        except AttributeErrir:
```
yowza

fonnesbeck commented:
Yes, go for it.

ColCarroll reviewed (commented):
can't figure out why it isn't building -- can you either make some of the suggested changes and push, or just `git commit --allow-empty` and then push?

ColCarroll commented on pymc3/distributions/multivariate.py:
```diff
@@ -372,8 +372,8 @@ def __init__(self, a, transform=transforms.stick_breaking,
-        self.k = tt.as_tensor_variable(shape)
-        self.a = a = tt.as_tensor_variable(a)
+        self.k = floatX(shape)
+        self.a = a = floatX(a)
```
total nit, and not necessary, but might be nicer to just use `self.a` below, and stop referring to `a` (same with `Sigma` and `nu`.

hvasbath commented:
The test failed, because you need scipy 0.18 otherwise there is no skewnorm. The requirements have to be updated!
Also I get tons of PEP8 errors too long lines... for now I ignored all of them, as you dont seem to care either?

hvasbath committed:
remove double assignments, intX casting, get name function

hvasbath committed:
remove double assignments, intX casting, get name function

hvasbath commented:
Travis still not starting ...?
Locally test_distributions passes now. However in MvStundentT there is strange behavior. There n should be an int by nature, but the test fails if it is not a float. l. 651. But I have not enough insight in the test and no time to investigate that quite complex structure. So I cast it to float and edit the latex test ...
Also I introduce an intX function. If we use float32 to run on the GPU all Ints should be cast to int16, because interaction of int32 and above with float32 results in float64, what we dont want...

junpenglao commented:
https://travis-ci.org/pymc-devs/pymc3/requests
```
abuse detected: known offender (request looked fishy) 
``` 
What have you done @hvasbath 😂!

hvasbath commented:
WTH???? I am sitting in Saudi Arabia-maybe my IP is suspicious ... :)
K thanks @junpenglao I did that lets see...

junpenglao commented:
You are on the naughty list :joy:!
You should email support@travis-ci.com https://github.com/travis-ci/travis-ci/issues/6942

hvasbath committed:
travis update

hvasbath commented:
Had to connect git with travis to resolve that- apparently this was a false positive of their malware detection ...

hvasbath committed:
updated username locally

hvasbath commented:
OMG all these failed tests! Where do all these float64 casts come from? I have no time right now to look into all of these tests and whats going wrong, some of them as well assume the variables to be tensors ... not sure what are your oppinions?

junpenglao commented:
I dont think it's that bad, you just need to recast some of the parameters into tensor.

hvasbath commented:
But this is the point in the first place! We do not want to do that in order to have the random method perform fast ...

Commit:
fix failed test

twiecki commented:
That is surprising, @hvasbath I would imagine that the a single issue underlies most of the failing tests, so perhaps just look into one and I would hope that it would resolve the other ones.

junpenglao commented:
> But this is the point in the first place! We do not want to do that in order to have the random method perform fast ...

I know, as long as they are not tensor in the `random` method right?
I fixed some test, still failing in 
```
pymc3/tests/sampler_fixtures.py:130: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/tests/sampler_fixtures.py:155: in make_step
    mu, stds, elbo = pm.advi(n=50000)
pymc3/variational/advi.py:164: in advi
    uw_i, elbo_current = f()
/usr/local/lib/python3.5/dist-packages/theano/compile/function_module.py:898: in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
/usr/local/lib/python3.5/dist-packages/theano/gof/link.py:325: in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
../../../.local/lib/python3.5/site-packages/six.py:685: in reraise
    raise value.with_traceback(tb)
```

hvasbath committed:
WIP continued floatX casting and rebase

hvasbath committed:
Merge branch 'random_slowdown' of https://github.com/hvasbath/pymc3 into random_slowdown

Conflicts:
	pymc3/distributions/discrete.py

hvasbath committed:
fixes tests in variational_inference

hvasbath commented:
Rebased the PR, fixed tests in variational_inference. 
Something like float32 -1 will result in float64! Please keep that in mind while coding ...I had to fix a lot of these things in the tests. I wonder why the tests didnt fail before? I guess it works with tensors as theano can detect this and autocasts?
Any comments? Maybe I got something totally wrong here...

hvasbath committed:
fix test_distributions_random

hvasbath commented:
This likely only needs the new fix from @junpenglao included, then also these last tests should pass. However, I cannot rebase again without getting hundreds of conflicts. Can anybody say why?
I tried in this branch
git rebase master
When I did it yesterday it gave only few conflicts that were resolvable by hand. However, now at every line that I changed here, it reports a conflict ...

hvasbath committed:
Merge branch 'master' into random_slowdown

hvasbath commented:
Again what learned :) .. Merge did the trick and all the mess went away.

junpenglao commented on pymc3/distributions/continuous.py:
```diff
@@ -824,21 +826,20 @@ class Pareto(PositiveContinuous):
+        self.alpha = floatX(alpha)
+        self.m = floatX(m)
 
         self.mean = tt.switch(tt.gt(alpha, 1), alpha *
```
shouldnt here be `self.alpha` instead of just `alpha`

hvasbath replied:
Yep missed that!

junpenglao commented on pymc3/distributions/discrete.py:
```diff
@@ -438,17 +441,17 @@ class DiscreteUniform(Discrete):
         super(DiscreteUniform, self).__init__(*args, **kwargs)
-        self.lower = tt.floor(lower).astype('int32')
-        self.upper = tt.floor(upper).astype('int32')
+        self.lower = intX(np.floor(lower))
```
Will it work if `lower` being a tensor? Is `intX(lower)` sufficient?

hvasbath replied:
Having theano tensor as distribution bound inputs? Is that intended? I should have known that in advance! If yes we can forget about this whole PR ...

junpenglao replied:
Yeah theano tensor could be distribution bound.

In this case can you not do something similar to `Uniform(Continuous)`:
```python
        self.lower = floatX(lower)
        self.upper = floatX(upper)
```

hvasbath commented:
Oh no due to the rebase and then the merge there are duplicated commits now! Argh ... How to fix that?

twiecki commented:
@hvasbath Looks OK to me. We will do a squash-merge in any case so it doesn't really matter.

hvasbath commented:
ok then I will close the other PR again.

hvasbath commented:
Continued in the other cleaned up PR!
